StrongInject.Tests.Unit/OwnedInjectionTests.cs

```diff
@@ -0,0 +1,1950 @@
+}");
+        }
+
+        [Fact]
+        public void OwnedCanBeManuallyProvidedViaFactory()
+        {
+            string userSource = @"
+using System;
+using StrongInject;
+
+[Register(typeof(A)), Register(typeof(B))]
+public partial class Container : IContainer<A>
+{
+    [Factory]
+    public static Owned<B> CreateOwnedManually() => throw new NotImplementedException();
+}
+
+public record A(Owned<B> b);
+public record B : IDisposable { public void Dispose() { } }
+";
+            var comp = RunGeneratorWithStrongInjectReference(userSource, out var generatorDiagnostics, out var generated);
+            generatorDiagnostics.Verify();
+            comp.GetDiagnostics().Verify();
+            var file = Assert.Single(generated);
+            file.Should().BeIgnoringLineEndings(@"#pragma warning disable CS1998
+partial class Container
+{
+    private int _disposed = 0;
+    private bool Disposed => _disposed != 0;
+    public void Dispose()
+    {
+        var disposed = global::System.Threading.Interlocked.Exchange(ref this._disposed, 1);
+        if (disposed != 0)
+            return;
+    }
+
+    TResult global::StrongInject.IContainer<global::A>.Run<TResult, TParam>(global::System.Func<global::A, TParam, TResult> func, TParam param)
+    {
+        if (Disposed)
+            throw new global::System.ObjectDisposedException(nameof(Container));
+        global::StrongInject.Owned<global::B> owned_0_1;
+        global::A a_0_0;
+        owned_0_1 = global::Container.CreateOwnedManually();
+        try
+        {
+            a_0_0 = new global::A(b: owned_0_1);
+        }
+        catch
+        {
+            ((global::System.IDisposable)owned_0_1).Dispose();
+            throw;
+        }
+
+        TResult result;
+        try
+        {
+            result = func(a_0_0, param);
+        }
+        finally
+        {
+            ((global::System.IDisposable)owned_0_1).Dispose();
+        }
+
+        return result;
+    }
+
+    global::StrongInject.Owned<global::A> global::StrongInject.IContainer<global::A>.Resolve()
+    {
+        if (Disposed)
+            throw new global::System.ObjectDisposedException(nameof(Container));
+        global::StrongInject.Owned<global::B> owned_0_1;
+        global::A a_0_0;
+        owned_0_1 = global::Container.CreateOwnedManually();
+        try
+        {
+            a_0_0 = new global::A(b: owned_0_1);
+        }
+        catch
+        {
+            ((global::System.IDisposable)owned_0_1).Dispose();
+            throw;
+        }
+
+        return new global::StrongInject.Owned<global::A>(a_0_0, () =>
+        {
+            ((global::System.IDisposable)owned_0_1).Dispose();
+        });
+    }
+}");
+        }
+
+        // Demonstrates another especially useless thing to do as currently producing working code with no StrongInject diagnostics.
+        // (There's no way to provide a semantically correct dispose action if the contained instance or any of its dependencies are disposable.)
```

Comment:
Re: https://github.com/YairHalberstadt/stronginject/pull/143#discussion_r698273545

> > Should also test how StrongInject should react to people explicitly trying to register Async/Owned.
> 
> My philosophy so far is that works, and it overrides the default provision of `Owned` - so basically `Owned` just becomes a normal registration with no special behaviour.

Are you sure you don't want to error in this situation? Is there any way that registering `Owned<>` explicitly will not be either subtly dangerous if not closely examined (if people are just trying to unblock themselves by doing this), or at least a path to go down that is a waste of time?

Reply:
I think you're right, but it might end up very involved, and it's not critical.

A couple of points:

We should allow registering a decorator for Owned<T>.
We should ideally handle the case where somebody registers a `[Factory] T GetAnything<T>` and not use it if `T` is an `Owned<T>`. I think the most sensible option for me would probably be to just disallow that completely, which would be a breaking change, but I realize that there's almost no way to use that safely. The reason I allowed it is because maybe you want a last chance provider which tries to create it via reflection, or by calling another container, but that just doesn't seem sensible. So no need for you to handle that yourself - I'm happy to do so.

Reply:
> I think you're right, but it might end up very involved, and it's not critical.

If I find a simple approach, are you happy with adding this as an error diagnostic?

> We should allow registering a decorator for Owned.

I can add a test for this and see I find.

> We should ideally handle the case where somebody registers a `[Factory] T GetAnything<T>` and not use it if `T` is an `Owned<T>`. I think the most sensible option for me would probably be to just disallow that completely, which would be a breaking change, but I realize that there's almost no way to use that safely.

This isn't what I expected. Why wouldn't there be a safe way to return `Owned<X>` or `AsyncOwned<X>` from a `[Factory]`?

Reply:
> If I find a simple approach, are you happy with adding this as an error diagnostic?

Sure but let's leave that for a different PR - it shouldn't hold this one up.